File: minesweeper.py
import random

N = -1
B = -1
board = []
finished = False
lose = False
# Constants
BOMB = -1


# Tile is the struct of a single tile in the board
class Tile:
    val = 0
    opened = False

# init_board initialize the game board
def init_board(N, B):
    global board
    board = [[Tile() for i in range(N+2)] for j in range(N+2)]

    bombs_planted = 0
    while (bombs_planted < B):
        randx, randy = random.randint(1, N), random.randint(1, N)
        if board[randx][randy].val != BOMB:
            board[randx][randy].val = BOMB
            bombs_planted += 1

    for i in range(1, N+1):
        for j in range(1, N+1):
            if board[i][j].val == BOMB:
                add_value(i-1, j+1)
                add_value(i, j+1)
                add_value(i+1, j+1)
                add_value(i-1, j)
                add_value(i+1, j)
                add_value(i-1, j-1)
                add_value(i, j-1)
                add_value(i+1, j-1)

    return


# add_value increase the counter of surrounding bombs of a tile
def add_value(x, y):
    global board
    if board[x][y].val != BOMB:
        board[x][y].val += 1
    return

# input_tile reads a user input and open it
def input_tile():
    x, y = input("Input Row , Column (eg. : '1 4' inputs the first row and the fourth column) : ").split()
    x, y = int(x), int(y)
    while x < 1 or y < 1 or x > N or y > N or board[x][y].opened == True:
        print('Invalid Row,Col')
        x, y = input("Input Row , Column (eg. : '1 4' inputs the first row and the fourth column) : ").split()
        x, y = int(x), int(y)
    return open_tile(x, y)

# open_tile opens a tile and stop the game if a Bomb is opened
def open_tile(x, y):
    global board
    if x < 1 or y < 1 or x > N or y > N or board[x][y].opened:
        return False
    if board[x][y].val == BOMB:
        return True

    board[x][y].opened = True

    if board[x][y].val == 0:
        open_tile(x-1, y-1)
        open_tile(x, y-1)
        open_tile(x+1, y-1)
        open_tile(x-1, y)
        open_tile(x+1, y)
        open_tile(x-1, y+1)
        open_tile(x, y+1)
        open_tile(x+1, y+1)

    return False

# print_board prints the game's board
def print_board():
    print()
    print('Remaining Bombs : ', B)
    print(end='   ')
    for i in range(1, N+1):
        print(i, end='  ')
        
    print()
    for i in range(1, N+1):
        print(i, end=' ')
        if i < 10:
            print(end=' ')
        for j in range(1, N+1):
            if not board[i][j].opened:
                print('#', end='  ')
            else:
                if board[i][j].val == BOMB:
                    print('X', end='  ')
                else:
                    print(board[i][j].val, end='  ')
        print()
    print()
    return

# print_end_game prints the true board which will be called at the end of the game
def print_end_game():
    global board
    for i in range(1, N+1):
        for j in range(1, N+1):
            board[i][j].opened = True
    print_board()
    return

# is_finished checks is the game has been won or not
def is_finished():
    tiles_opened = 0
    for i in range(1, N+1):
        for j in range(1, N+1):
            if board[i][j].opened:
                tiles_opened += 1
    return B == (N**2)-tiles_opened


if __name__ == "__main__":
    print('========== Minesweeper by Cornelius Yan M. - 13516113 ==========')
    N = int(input('Input N : '))
    B = int(input('Input number of Bombs : '))
    while B >= N**2 or B < 1:
        print('Input Bombs invalid ! Bombs must be around 0 < B < ', N**2)
        B = int(input('Input number of Bombs : '))

    init_board(N, B)

    while not finished:
        print_board()
        lose = input_tile()
        if not lose:
            finished = is_finished()
        else:
            finished = True

    print()
    if not lose:
        print('========== Game Over! You Win. ==========')
    else:
        print('========== Game Over! You lose. ==========')
    print_end_game()

File: Minesweeper.cpp
 
#include<bits/stdc++.h>
using namespace std;

#define MAXSIDE 25
#define MAXMINES 99
#define MOVESIZE 526 // (25 * 25 - 99)

int SIDE ; // side length of the board
int MINES ; // number of mines on the board
 
// A Utility Function to check whether given cell (row, col)
// is a valid cell or not
bool isValid(int row, int col)
{
    // Returns true if row number and column number
    // is in range
    return (row >= 0) && (row < SIDE) &&
           (col >= 0) && (col < SIDE);
}
 
// A Utility Function to check whether given cell (row, col)
// has a mine or not.
bool isMine (int row, int col, char board[][MAXSIDE])
{
    return (board[row][col] == '*');
}
 
// A Function to get the user's move
void makeMove(int *x, int *y)
{
    // Take the input move
    printf(">> Enter your move, (row, column) -> ");
    scanf("%d %d", x, y);
}
 
// A Function to print the current gameplay board
void printBoard(char myBoard[][MAXSIDE])
{
    int i, j;
 
    printf ("    ");
 
    for (i=0; i<SIDE; i++)
        printf ("%d ", i);
 
    printf ("\n\n");
 
    for (i=0; i<SIDE; i++)
    {
        printf ("%d   ", i);
 
        for (j=0; j<SIDE; j++)
            printf ("%c ", myBoard[i][j]);
        printf ("\n");
    }
    return;
}
 
// A Function to count the number of
// mines in the adjacent cells
int countAdjacentMines(int row, int col, int mines[][2], char realBoard[][MAXSIDE]){
 
    int i;
    int count = 0;
 
    if (isValid (row-1, col)){
        if (isMine (row-1, col, realBoard))
            count++;
    }
 
    if (isValid (row+1, col)){
        if (isMine (row+1, col, realBoard))
            count++;
    }
 
    if (isValid (row, col+1)){
        if (isMine (row, col+1, realBoard))
            count++;
    }
 
    if (isValid (row, col-1)){
        if (isMine (row, col-1, realBoard))
            count++;
    }
 
    if (isValid (row-1, col+1)){
        if (isMine (row-1, col+1, realBoard))
            count++;
    }
 
    if (isValid (row-1, col-1)){
        if (isMine (row-1, col-1, realBoard))
            count++;
    }
 
    if (isValid (row+1, col+1)){
        if (isMine (row+1, col+1, realBoard))
            count++;
    }
 
    if (isValid (row+1, col-1)){
        if (isMine (row+1, col-1, realBoard))
            count++;
    }
 
    return count;
}
 
//A Recursive Fucntion to play the Minesweeper Game
bool continuePlay(char myBoard[][MAXSIDE], char realBoard[][MAXSIDE], int mines[][2], int row, int col, int *movesLeft){
 
    // Base Case
    if (myBoard[row][col] != '-'){
        return false;
    }
 
    int i, j;
 
    // a mine opened
    // going to lose
    if (realBoard[row][col] == '*'){
        myBoard[row][col]='*';
 
        for (i=0; i<MINES; i++)
            myBoard[mines[i][0]][mines[i][1]]='*';
 
        printBoard (myBoard);
        printf ("\nYou lost!\n");
        return true;
    }
 
    else
     {
        // Calculate the number of adjacent mines and put it
        // on the board
        int count = countAdjacentMines(row, col, mines, realBoard);
        (*movesLeft)--;
 
        myBoard[row][col] = count + '0';
 
        if (count==0){
            if (isValid (row-1, col)){
                if (!isMine (row-1, col, realBoard))
                    continuePlay(myBoard, realBoard, mines, row-1, col, movesLeft);
            }
 
            if (isValid (row+1, col)){
                if (!isMine (row+1, col, realBoard))
                    continuePlay(myBoard, realBoard, mines, row+1, col, movesLeft);
            }
 
            if (isValid (row, col+1)){
                if (!isMine (row, col+1, realBoard))
                    continuePlay(myBoard, realBoard, mines, row, col+1, movesLeft);
            }
 
            if (isValid (row, col-1)){
                   if (!isMine (row, col-1, realBoard))
                    continuePlay(myBoard, realBoard, mines, row, col-1, movesLeft);
            }
 
            if (isValid (row-1, col+1)){
                if (!isMine (row-1, col+1, realBoard))
                    continuePlay(myBoard, realBoard, mines, row-1, col+1, movesLeft);
            }
 
            if (isValid (row-1, col-1)){
                 if (!isMine (row-1, col-1, realBoard))
                    continuePlay(myBoard, realBoard, mines, row-1, col-1, movesLeft);
            }
 
            if (isValid (row+1, col+1)){
                 if (!isMine (row+1, col+1, realBoard))
                    continuePlay(myBoard, realBoard, mines, row+1, col+1, movesLeft);
            }

            if (isValid (row+1, col-1)){
                if (!isMine (row+1, col-1, realBoard))
                    continuePlay(myBoard, realBoard, mines, row+1, col-1, movesLeft);
            }
        }
 
        return false;
    }
}
 
// A Function to place the mines randomly
// on the board
void placeMines(int mines[][2], char realBoard[][MAXSIDE]){
    bool mem[MAXSIDE*MAXSIDE];
 
    memset (mem, false, sizeof (mem));
 
    // Continue until all random mines have been created.
    for (int i=0; i<MINES; )
     {
        int random = rand() % (SIDE*SIDE);
        int x = random / SIDE;
        int y = random % SIDE;
 
        // Add the mine if no mine is placed at this
        // position on the board
        if (mem[random] == false)
        {
            // Row Index of the Mine
            mines[i][0]= x;
            // Column Index of the Mine
            mines[i][1] = y;
 
            // Place the mine
            realBoard[mines[i][0]][mines[i][1]] = '*';
            mem[random] = true;
            i++;
        }
    }
}
 
// A Function to initialise the game
void initialise(char realBoard[][MAXSIDE], char myBoard[][MAXSIDE]){
    // Initiate the random number generator so that
    // the same configuration doesn't arises
    srand(time (NULL));
 
    // Assign all the cells as mine-free
    for (int i=0; i<SIDE; i++)
    {
        for (int j=0; j<SIDE; j++)
        {
            myBoard[i][j] = realBoard[i][j] = '-';
        }
    }
}
 
// A Function to cheat by revealing where the mines are
// placed.
void cheatMinesweeper (char realBoard[][MAXSIDE]){
    printf ("The mines locations are-\n");
    printBoard (realBoard);
    return;
}
 
// A function to replace the mine from (row, col) and put
// it to a vacant space
void replaceMine (int row, int col, char board[][MAXSIDE]){
    for (int i=0; i<SIDE; i++){
        for (int j=0; j<SIDE; j++){
            if (board[i][j] != '*'){
                board[i][j] = '*';
                board[row][col] = '-';
                return;
            }
        }
    }
}
 
// A Function to play Minesweeper game
void playMinesweeper (){
    // Initially the game is not over
    bool gameOver = false;
 
    // Actual Board and My Board
    char realBoard[MAXSIDE][MAXSIDE], myBoard[MAXSIDE][MAXSIDE];
 
    int movesLeft = SIDE * SIDE - MINES;
    int x, y;
    int mines[MAXMINES][2]; // stores (x,y) coordinates of all mines.
 
    initialise (realBoard, myBoard);
 
    // Place the Mines randomly
    placeMines (mines, realBoard);
 
     /*
     If you want to cheat and know
     where mines are before playing the game
     then uncomment this part
    */
    //cheatMinesweeper(realBoard);
 
    // You are in the game until you have not opened a mine
    // So keep playing
 
    int currentMoveIndex = 0;
    while (gameOver == false){
        printf ("Current Status of Board : \n");
        printBoard (myBoard);
        makeMove (&x, &y);
 
        if (currentMoveIndex == 0){
            // If the first move itself is a mine
            // then we remove the mine from that location
            if (isMine (x, y, realBoard))
                replaceMine (x, y, realBoard);
        }
 
        currentMoveIndex ++;
 
        gameOver = continuePlay (myBoard, realBoard, mines, x, y, &movesLeft);
 
        if ((!gameOver) && (movesLeft == 0)){
            printf ("\nYou won !\n");
            gameOver = true;
        }
    }
}
 
// A Function to choose the difficulty level
// of the game
void startGame (){
    int N, B;
    bool valid = false;                                                                                                                   

    printf("             . . .\n");                         
    printf("              \\|/\n");                          
    printf("            `--+--'\n");                        
    printf("              /|\\\n");                          
    printf("             ' | '\n");                         
    printf("               |\n");                           
    printf("               |\n");                           
    printf("           ,--'#`--.\n");                       
    printf("           |#######|\n");                       
    printf("        _.-'#######`-._\n");                    
    printf("     ,-'###############`-.\n");                 
    printf("   ,'#####################`,\n");               
    printf("  /#########################\\\n");              
    printf(" |###########################|\n");             
    printf("|#############################|\n");            
    printf("|#############################|\n");            
    printf("|#############################|\n");            
    printf("|#############################|\n");            
    printf(" |###########################|\n");             
    printf("  \\#########################/\n");              
    printf("   `.#####################,'\n");               
    printf("     `._###############_,'\n");                 
    printf("        `--..#####..--'\n");
    printf("\n");
    printf("             BOOM !\n");
    printf("\n");
    printf ("--- Welcome to Minesweeper Console ---\n");
    printf ("      brought to you by 13516081      \n");
    
    while(!valid){
        printf (">> Enter the board size -> ");
        scanf ("%d", &N);
        printf (">> Enter the number of Mines -> ");
        scanf ("%d", &B);
        if(B <= N*N){
            valid = true;
        }else{
            printf(">> Your input is invalid\n");
        }
    }
    SIDE = N;
    MINES = B;
}
 
// Driver Program to test above functions
int main(){
    startGame();
 
    playMinesweeper ();
 
    return 0;
}